application/ledger_query_service/src/ethereum/queries.rs

```diff
@@ -148,45 +179,83 @@ impl IsEmpty for TransactionQuery {
     }
 }
 
-impl LogQuery {
+impl EventQuery {
     pub fn matches_block(&self, block: &Block<Transaction>) -> bool {
         match self {
-            Self { topics, .. } if topics.is_empty() => false,
-            Self { topics } => topics.iter().all(|topics| {
-                topics
-                    .iter()
-                    .all(|topic| block.logs_bloom.contains_input(Input::Raw(&topic)))
+            Self { event_matchers } if event_matchers.is_empty() => false,
+            Self { event_matchers } => event_matchers.iter().all(|log_matcher| {
+                log_matcher.topics.iter().all(|topic| {
+                    if let Some(topic) = topic {
+                        block.logs_bloom.contains_input(Input::Raw(&topic))
+                    } else {
+                        false
+                    }
+                })
             }),
         }
     }
 
     pub fn matches_transaction_receipt(&self, transaction_receipt: TransactionReceipt) -> bool {
         match self {
-            Self { topics } if topics.is_empty() => false,
-            Self { topics } => topics.iter().all(|topics| {
-                !topics.is_empty()
-                    && transaction_receipt
-                        .logs
-                        .iter()
-                        .any(|tx_log| topics.iter().all(|topic| tx_log.topics.contains(topic)))
+            Self { event_matchers } if event_matchers.is_empty() => false,
+            Self { event_matchers } => event_matchers.iter().all(|event_matcher| {
+                return match event_matcher {
+                    EventMatcher {
+                        address: None,
+                        data: None,
+                        topics: topics,
+                    } if topics.is_empty() => false,
+                    EventMatcher {
+                        address: address,
+                        data: data,
+                        topics: topics,
+                    } => transaction_receipt.logs.iter().any(|tx_log| {
+                        let mut result = true;
+
+                        if let Some(address) = address {
+                            result = result && (address == &tx_log.address);
```

Comment:
Early return would be cleaner IMO

Reply:
Mhmm... that's not what we want. We want to check all required filters if they are present.